Composer/plugins/lib/bot-deploy/src/utils/crossTrainUtil.ts

```diff
@@ -0,0 +1,215 @@
+    const fileNames = getLuFilesByDialogId(key, luFiles);
+    return {
+      ...result,
+      ...fileNames.reduce((result, name) => {
+        const locale = getFileLocale(name);
+        const triggers = triggerRules[key];
+        keys(triggers).forEach((trigger) => {
+          if (!result[name]) result[name] = {};
+          const ids = triggers[trigger];
+          if (Array.isArray(ids)) {
+            result[name][trigger] = ids.map((id) => (id ? `${id}.${locale}.lu` : id));
+          } else {
+            result[name][trigger] = ids ? `${ids}.${locale}.lu` : ids;
+          }
+        });
+        return result;
+      }, {}),
+    };
+  }, {});
+  return config;
+}
+
+function parse(dialog) {
+  const { id, content, isRoot } = dialog;
+  const luFile = typeof content.recognizer === 'string' ? getBaseName(id) : '';
+  const qnaFile = typeof content.recognizer === 'string' ? getBaseName(id) : '';
+
+  return {
+    id: getBaseName(id),
+    isRoot: isRoot,
+    content,
+    luFile: luFile,
+    qnaFile: qnaFile,
+    intentTriggers: ExtractIntentTriggers(content),
+  };
+}
+export interface ICrossTrainConfig {
+  rootIds: string[];
+  triggerRules: { [key: string]: any };
+  intentName: string;
+  verbose: boolean;
+  botName: string;
+}
+
+//generate the cross-train config without locale
+/* the config is like
+  {
+      rootIds: [
+        'main.en-us.lu',
+        'main.fr-fr.lu'
+      ],
+      triggerRules: {
+        'main.en-us.lu': {
+          'dia1_trigger': 'dia1.en-us.lu',
+          'dia2_trigger': 'dia2.en-us.lu'
+        },
+        'dia2.en-us.lu': {
+          'dia3_trigger': 'dia3.en-us.lu',
+          'dia4_trigger': 'dia4.en-us.lu'
+        },
+        'main.fr-fr.lu': {
+          'dia1_trigger': 'dia1.fr-fr.lu'
+        }
+      },
+      intentName: '_Interruption',
+      verbose: true
+    }
+  */
+export function createCrossTrainConfig(dialogs: any[], luFilesInfo: FileInfo[]): ICrossTrainConfig {
+  const triggerRules = {};
+  const countMap = {};
+  const wrapDialogs: { [key: string]: any }[] = [];
+  for (const dialog of dialogs) {
+    wrapDialogs.push(parse(dialog));
+  }
+
+  const luFiles = luIndexer.index(luFilesInfo);
+
+  //map all referred lu files
+  luFiles.forEach((file) => {
+    countMap[getBaseName(file.id)] = 1;
+  });
+
+  let rootId = '';
+  let botName = '';
+  wrapDialogs.forEach((dialog) => {
+    if (dialog.isRoot) {
+      rootId = dialog.id;
+      botName = dialog.content.$designer.name;
+    }
+
+    if (luFiles.find((luFile) => getBaseName(luFile.id) === dialog.luFile)) {
+      const { intentTriggers } = dialog;
+      const fileId = dialog.id;
+      //find the trigger's dialog that use a recognizer
+      intentTriggers.forEach((item) => {
+        //find all dialogs in trigger that has a luis recognizer
+        const used = item.dialogs.filter((dialog) => !!countMap[dialog]);
+
+        const deduped = Array.from(new Set<string>(used));
```

Comment:
How it's related to `luUtils.ts`? Code is the same

Reply:
It is the same. Copy the logic is because we can not use client logic in plugin.